eggtimer.js

```diff
@@ -1,326 +1,681 @@
+            if (currentContext === null)
+                processNextPR(true);
+        });
+}
+
+function signaled() {
+    return (currentContext !== null && currentContext.signaled);
+}
+
+function processEvent() {
+    if (currentContext !== null) {
+        console.log("Signaling PR" + currentContext.pr.number, currentContext.pr.head.sha);
+        currentContext.signaled = true;
+    } else {
+        processNextPR(true);
+    }
+}
+
+function processNextPR(all) {
+    const sig = signaled();
+    currentContext = null;
+    if (all) {
+        getPRList();
+        return;
+    } else if (PRList.length === 0 && !sig) {
+        console.log("No more PRs to process.");
+        return;
+    } else if (sig) {
+        getPRList();
+        return;
     }
+    let prContext = createPRContext(PRList.shift());
+    console.log("Processing PR" + prContext.pr.number, prContext.pr.head.sha);
+    currentContext = prContext;
+
+    checkMergePreconditions(prContext);
+}
+
+function mergingTag(prNum) {
+    return "tags/" + MergingTag + prNum;
+}
+
+function checkMergePreconditions(prContext) {
+    let reviewsParams = commonParams();
+    reviewsParams.number = prContext.pr.number;
+    let reviewsPromise = getReviews(reviewsParams);
+
+    let statusParams = commonParams();
+    statusParams.ref = prContext.pr.head.sha;
+    let statusPromise = getStatuses(statusParams);
+
+    let mergeableParams = commonParams();
+    mergeableParams.number = prContext.pr.number;
+    let mergeablePromise = getPR(mergeableParams);
+
+    let labelsParams = commonParams();
+    labelsParams.number = prContext.pr.number;
+    let labelsPromise = getLabels(labelsParams);
+
+    Promise.all([reviewsPromise, statusPromise, mergeablePromise, labelsPromise]).then((results) => {
+            if (prContext.signaled)
+                return Promise.reject("signaled");
+
+            let reviews = results[0];
+            let statuses = results[1];
+            let mergeable = results[2].mergeable;
+            let labels = results[3];
+
+            if ((mergeable !== true) || !approved(reviews) || !allChecksSuccessful(statuses))
+                return Promise.reject("not ready for merging yet");
+
+            if (markedAsMerged(labels))
+                return Promise.reject("already merged");
+
+            if (markedAsFailed(labels))
+                console.log("PR" + prContext.pr.number + " previous merge was unsuccessful");
+
+            checkBeforeStartMerging(prContext);
+            return Promise.resolve(true);
+    })
+    .catch((err) => {
+        console.error("Can't merge PR" + prContext.pr.number + ":", err);
+        processNextPR();
+    });
+}
+
+function checkBeforeStartMerging(prContext) {
+   let refParams = commonParams();
+   refParams.ref = mergingTag(prContext.pr.number);
+   let tagTreeSha = null;
+   prContext.tagSha = null;
+   console.log("Checking whether to skip due to a previous unsuccessful merge for PR" + prContext.pr.number);
```

Comment:
The code below appears to be doing more than that. For example it also checks test statuses. In fact, it is difficult for me to find the place where we "skip due to a previous unsuccessful merge".

Reply:
I [made](https://github.com/measurement-factory/EggTimer/pull/1#discussion_r153134008) a quick explanation of 'checkBeforeStartMerging()' purpose. Yes, it checks the 'merging' commit statuses(which could have been changed) to make the appropriate decision.

Reply:
Please note that explaining checkBeforeStartMerging() purpose does nothing to address my concern, even if I completely agree with the explained purpose: When I or somebody else looks at this code a year from now, they will have the same concern.